core/blockchain.go

```diff
@@ -2135,6 +2176,73 @@ func (bc *BlockChain) ReadShardLastCrossLink(shardID uint32) (*types.CrossLink,
 	return crossLink, err
 }
 
+// ReadPendingCrossLinks retrieves pending crosslinks
+func (bc *BlockChain) ReadPendingCrossLinks() ([]types.CrossLink, error) {
+	bytes := []byte{}
+	if cached, ok := bc.pendingCrossLinksCache.Get("pendingCLs"); ok {
+		bytes = cached.([]byte)
+	} else {
+		bytes, err := rawdb.ReadPendingCrossLinks(bc.db)
+		if err != nil || len(bytes) == 0 {
+			utils.Logger().Info().Err(err).Int("dataLen", len(bytes)).Msg("ReadPendingCrossLinks")
+			return nil, err
+		}
+	}
+	cls := []types.CrossLink{}
+	if err := rlp.DecodeBytes(bytes, &cls); err != nil {
+		utils.Logger().Error().Err(err).Msg("Invalid pending crosslink RLP decoding")
+		return nil, err
+	}
+	return cls, nil
+}
+
+// WritePendingCrossLinks saves the pending crosslinks
+func (bc *BlockChain) WritePendingCrossLinks(cls []types.CrossLink) error {
+	bytes, err := rlp.EncodeToBytes(cls)
+	if err != nil {
+		utils.Logger().Error().Msg("[WritePendingCrossLinks] Failed to encode pending crosslinks")
+		return err
+	}
+	if err := rawdb.WritePendingCrossLinks(bc.db, bytes); err != nil {
+		return err
+	}
+	by, err := rlp.EncodeToBytes(cls)
+	if err == nil {
+		bc.pendingCrossLinksCache.Add("pendingCLs", by)
+	}
+	return nil
+}
+
+// AddPendingCrossLinks appends pending crosslinks
+func (bc *BlockChain) AddPendingCrossLinks(pendingCLs []types.CrossLink) (int, error) {
+	cls, err := bc.ReadPendingCrossLinks()
+	if err != nil || len(cls) == 0 {
+		err := bc.WritePendingCrossLinks(pendingCLs)
+		return len(pendingCLs), err
+	}
+	cls = append(cls, pendingCLs...)
+	err = bc.WritePendingCrossLinks(cls)
+	return len(cls), err
+}
+
+// DeleteCommittedFromPendingCrossLinks delete pending crosslinks that already committed
+func (bc *BlockChain) DeleteCommittedFromPendingCrossLinks() (int, error) {
+	cls, err := bc.ReadPendingCrossLinks()
+	if err != nil || len(cls) == 0 {
+		return 0, err
+	}
+	pendingCLs := []types.CrossLink{}
+	for _, cl := range cls {
+		cl0, err := bc.ReadCrossLink(cl.ShardID(), cl.BlockNum())
```

Comment:
Let feed the crosslink from the block directly here instead of reading it from DB. Reading db is slow and it will further increase the risk of race condition (as we didn't lock on this method).